libs/entitlement-runtime/src/main/java/org/elasticsearch/entitlement/policy/EntitlementBuilder.java

```diff
@@ -0,0 +1,29 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the "Elastic License
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+package org.elasticsearch.entitlement.policy;
+
+import org.elasticsearch.xcontent.XContentParser;
+
+import java.io.IOException;
+
+public abstract class EntitlementBuilder {
+
+    protected final String policyName;
+    protected final String scopeName;
+    protected final XContentParser policyParser;
+
+    protected EntitlementBuilder(String policyName, String scopeName, XContentParser policyParser) {
+        this.policyName = policyName;
+        this.scopeName = scopeName;
+        this.policyParser = policyParser;
+    }
+
+    protected abstract void parseEntitlement() throws IOException;
```

Comment:
What is the contract of this method? From the implementations, it seems to use the parser to validate the structure, but doesn't actually have any other side effects or return anything?

Is the idea that this will ultimately return whatever data structure we end up with?

Reply:
I was thinking there would be two passes. One for parsing and one for building. The parsing one would validate things like is the structure correct and are these values strings, etc. The building would validate that the scopes exist and permissions have valid paths, etc then turn them into immutable structures that will be used for comparison. Open here to a single pass if we think that's better. Either way we will need to do a string lookup at some point to get the module object for comparison.

Reply:
I see. It makes sense, if you think there will be a **public** build method.
Maybe to make this clear these classes could be internal (package private?)

Reply:
That sounds like an interesting separation, between syntax and semantics. Phase 1: make sure you can tell what they're asking for; phase 2: make sure they're asking for something reasonable.

Reply:
I'm not 100% sure this is the right approach, yet, but have to start somewhere. Either way parsing code will be useful.